### PR TITLE
Removing MAC address random generation

### LINKED ISSUE #2943: Fix loose ends of GET /vm/config post-restore
This issue is a follow up of https://github.com/firecracker-microvm/firecracker/issues/2783.
The bulk of the issue was solved in this PR: https://github.com/firecracker-microvm/firecracker/pull/2806

There are still a couple of loose ends that could be fixed in order to get the best user experience:

- net devices MAC addresses are reported as all zeros if left unconfigured. View https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r793271297
- the `machine_config.smt` and `machine_config.cpu_templates` and `boot_source` values are lost (we could persist them in the `VmState`).

### PR DESCRIPTION
**Pull request by ShadowCurse**
Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

This PR:
- Removes random MAC address generation introduced by [this](https://github.com/firecracker-microvm/firecracker/pull/3121) commit
- fixes an issue in `GET /vm/config` which returned invalid `00:00:00:00:00:00` MAC address if it was not previously specified by the user.

Additionally fixes snapshot files not being truncated when saved.

## Reason
Fixes #2943 second time

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] New `unsafe` code is documented.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md

### CODE REVIEW

File: src/devices/src/virtio/net/persist.rs [hunk 1]
```diff
@@ -23,10 +24,39 @@ use super::{NUM_QUEUES, QUEUE_SIZE};
...
 
-#[derive(Clone, Versionize)]
+#[derive(Debug, Default, Clone, Versionize)]
 // NOTICE: Any changes to this structure require a snapshot version bump.
 pub struct NetConfigSpaceState {
-    guest_mac: [u8; MAC_ADDR_LEN],
+    #[version(end = 2, default_fn = "def_net_config_old")]
+    guest_mac_old: [u8; utils::net::mac::MAC_ADDR_LEN],
+    #[version(start = 2, de_fn = "de_net_config", ser_fn = "ser_net_config")]
+    guest_mac: Option<MacAddr>,
```
Comment (dianpopa) [inline src/devices/src/virtio/net/persist.rs:33]: Why not `Option<[u8; utils::net::mac::MAC_ADDR_LEN]>`, I am asking because this way it would not be necessary to make the utils crate dependent on Versionize

Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs:33]: We have `MacAddr` as type that represents MAC address and basically just a wrapper around `[u8; MAC_ADDR_LEN]`. 
I think it is better to use it, then just an array of bytes. It makes code easier to understand, without a bunch of `MacAddr::from_bytes_unchecked` calls.
(On this note I probably should make `guest_mac_old` also be `MacAddr`)

**Review by luminitavoicu — COMMENTED** [review]
To make this easier to review, how about we have a commit where you undo changes in #3121 and another commit with fixes for the mac returned in `vm/config`?

File: src/devices/src/virtio/net/device.rs [hunk 2]
```diff
@@ -172,18 +150,12 @@ impl Net {
...
-            } else {
-                // When the MAC address is not specified explicitly, generate one and assign it.
-                generate_mac_address()
-            }
-        };
-        config_space.guest_mac.copy_from_slice(mac_addr.get_bytes());
+        if let Some(mac) = guest_mac {
+            config_space.guest_mac = mac;
+            // When this feature isn't available, the driver generates a random MAC address.
+            // Otherwise, it should attempt to read the device MAC address from the config space.
```
Comment (vibharya) [inline src/devices/src/virtio/net/device.rs:156]: Can we reword the below comment better. A suggestion would be:
```suggestion
            // Enable feature to configure a MAC address
            // If not enabled, the driver generates an address
```

Comment (ShadowCurse) [reply src/devices/src/virtio/net/device.rs]: Updated

File: src/devices/src/virtio/net/persist.rs [hunk 3]
```diff
@@ -47,7 +47,7 @@ impl NetConfigSpaceState {
         // v1.1 and older versions do not have optional MAC address.
         warn!("Saving to older snapshot version, optional MAC address will not be saved.");
         match self.guest_mac {
-            Some(mac) => self.guest_mac_old = mac.into(),
```
Comment (pb8o) [inline src/devices/src/virtio/net/persist.rs]: We have a (soft) requirement that each commit should build and pass the tests. If this (and the others) fixes a clippy warning, it should go in the previous commit.

File: tests/integration_tests/performance/test_snapshot_perf.py [hunk 4]
```diff
@@ -118,8 +117,8 @@
                     "2vcpu_512mb.json": {"target": 60},
                 },
                 "async": {
-                    "2vcpu_256mb.json": {"target": 190},
```
Comment (pb8o) [inline tests/integration_tests/performance/test_snapshot_perf.py]: This change seems unrelated to the MAC address? Or does the snapshot latency performance change with this commit?

File: src/devices/src/virtio/net/device.rs [hunk 5]
```diff
@@ -152,8 +152,8 @@ impl Net {
         let mut config_space = ConfigSpace::default();
         if let Some(mac) = guest_mac {
             config_space.guest_mac = mac;
-            // When this feature isn't available, the driver generates a random MAC address.
```
Comment (pb8o) [inline src/devices/src/virtio/net/device.rs]: nit grammar: "will generate" or "generates"

File: src/utils/src/net/mac.rs [hunk 6]
```diff
@@ -104,6 +118,7 @@ impl MacAddr {
     /// let mac = MacAddr::from_bytes(&[0x01, 0x02, 0x03, 0x04, 0x05, 0x06]).unwrap();
     /// println!("{}", mac.to_string());
     /// ```
+    /// TODO this is unused, do we need it?
```
Comment (pb8o) [inline src/utils/src/net/mac.rs:121]: Good catch. I would say remove it as a separate commit.

File: src/vmm/src/persist.rs [hunk 7]
```diff
@@ -174,73 +173,47 @@ pub enum MicrovmStateError {
...
     /// Failed to open the snapshot backing file.
+    #[error("Cannot perform {0} on the snapshot backing file: {1}")]
     SnapshotBackingFile(&'static str, io::Error),
-    #[cfg(target_arch = "x86_64")]
     /// Number of devices exceeds the maximum supported devices for the snapshot data version.
+    #[cfg(target_arch = "x86_64")]
+    #[error(
+        "Too many devices attached: {0}. The maximum number allowed for the snapshot data version \
+         requested is {FC_V0_23_MAX_DEVICES}."
+    )]
```
Comment (pb8o) [inline src/vmm/src/persist.rs:213]: This seems unrelated to the MAC change. We have a PR best practice to have one logical change per commit. Can this be split to a separate commit?

**Comment by pb8o** [comment]
> > To make this easier to review, how about we have a commit where you undo changes in #3121 and another commit with fixes for the mac returned in `vm/config`?
> 
> @luminitavoicu I don't think I can `unsquash` commits. In addition, as I remember I did both changes at the same time, so they were in one commit anyway.

You can [unsquash](1) and reorder commits with `git rebase`. What I would do here: 

1. Start from a clean branch
2. `git revert` the relevant commits from #3121 
3. Cherry pick your 2 commits from this branch `git cherry-pick ...`
4. Do any other history manipulation as needed (merge commits, split commits) as mentioned in the comments
 
[1]: https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

File: docs/network-setup.md [hunk 8]
```diff
@@ -76,10 +76,6 @@ to your configuration file like this:
 Alternatively, if you are using firectl, add
 --tap-device=tap0/AA:FC:00:00:00:01` to your command line.
 
-*Note:* When `guest_mac` is not explicitely specified, a MAC address is generated
-and assigned to the network interface. This address is included in the result of
-GET `/vm/config`.
-
```
Comment (vibharya) [inline docs/network-setup.md:82]: Do we need to add any information regarding what happens when the user does not specify a MAC address?

```
*Note:* When `guest_mac` is not explicitly specified, a MAC address is generated
by the driver and assigned to the network interface. This address cannot be
queried via Firecracker
```

WDYT @dianpopa  @pb8o ?

File: tests/integration_tests/performance/test_snapshot_perf.py (hunk 4 above)
Comment (ShadowCurse) [reply tests/integration_tests/performance/test_snapshot_perf.py]: I am not sure why this is in one of the commits, this change is not shown in the `file changes`

**Comment by ShadowCurse** [comment]
> > > To make this easier to review, how about we have a commit where you undo changes in #3121 and another commit with fixes for the mac returned in `vm/config`?
> > 
> > 
> > @luminitavoicu I don't think I can `unsquash` commits. In addition, as I remember I did both changes at the same time, so they were in one commit anyway.
> 
> You can [unsquash](1) and reorder commits with `git rebase`. What I would do here:
> 
>     1. Start from a clean branch
> 
>     2. `git revert` the relevant commits from [Snapshot-ga: Generate a MAC address if unspecified #3121](https://github.com/firecracker-microvm/firecracker/pull/3121)
> 
>     3. Cherry pick your 2 commits from this branch `git cherry-pick ...`
> 
>     4. Do any other history manipulation as needed (merge commits, split commits) as mentioned in the comments

I did both changes at the same time, so I don't think this is `unsquash`able

**Comment by vibharya** [comment]
In addition to my other review comments, I believe changes need to be made in CHANGELOG.md as well.

Remove [this point](https://github.com/firecracker-microvm/firecracker/blob/main/CHANGELOG.md#:~:text=A%20MAC%20address%20is%20generated%20if%20one%20is%20not%20explicitly%20specified%20while%20adding%20network%20interfaces.%20This%20address%20can%20be%20obtained%20as%20part%20of%20the%20GET%20/vm/config.) and edit [this one](https://github.com/firecracker-microvm/firecracker/blob/main/CHANGELOG.md#:~:text=When%20MAC%20address%20was%20left%20unspecified%2C%20GET%20/vm/config%20would%20report%20the%20interface%20MAC%20address%20incorrectly%20as%20null%2C%20and%20post%20snapshot%20restore%20as%2000%3A00%3A00%3A00%3A00%3A00.%20It%20now%20correctly%20reports%20the%20code%20generated%20MAC%20address%20when%20queried%20in%20both%20cases.)

**Comment by vibharya** [comment]
In addition to my other review comments, I believe changes need to be made in CHANGELOG.md as well.

Remove [this point](https://github.com/firecracker-microvm/firecracker/blob/main/CHANGELOG.md#:~:text=A%20MAC%20address%20is%20generated%20if%20one%20is%20not%20explicitly%20specified%20while%20adding%20network%20interfaces.%20This%20address%20can%20be%20obtained%20as%20part%20of%20the%20GET%20/vm/config.) and edit [this one](https://github.com/firecracker-microvm/firecracker/blob/main/CHANGELOG.md#:~:text=When%20MAC%20address%20was%20left%20unspecified%2C%20GET%20/vm/config%20would%20report%20the%20interface%20MAC%20address%20incorrectly%20as%20null%2C%20and%20post%20snapshot%20restore%20as%2000%3A00%3A00%3A00%3A00%3A00.%20It%20now%20correctly%20reports%20the%20code%20generated%20MAC%20address%20when%20queried%20in%20both%20cases.)

File: src/devices/src/virtio/net/persist.rs [hunk 9]
```diff
@@ -67,7 +97,11 @@ impl Persist<'_> for Net {
             tx_rate_limiter_state: self.tx_rate_limiter.save(),
             mmds_ns: self.mmds_ns.as_ref().map(|mmds| mmds.save()),
             config_space: NetConfigSpaceState {
-                guest_mac: self.config_space.guest_mac,
+                guest_mac: self.guest_mac,
+                guest_mac_old: match self.guest_mac {
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs:101]: yup, seems like useless duplication, I will remove it

File: src/utils/src/net/mac.rs (hunk 6 above)
Comment (ShadowCurse) [reply src/utils/src/net/mac.rs:121]: If we don't want to check the length, then we have a `from_bytes_unchecked`

Comment (vibharya) [reply src/utils/src/net/mac.rs:121]: Yes I understand that there's another function.
I'm trying to make sure that we have no use case where we'll need to verify length. 

If there are none, we can remove it.

File: src/devices/src/virtio/net/device.rs (hunk 5 above)
Comment (ShadowCurse) [reply src/devices/src/virtio/net/device.rs]: Am I blind or there is not grammar error?

File: src/utils/src/net/mac.rs (hunk 6 above)
Comment (ShadowCurse) [reply src/utils/src/net/mac.rs]: well if we need to verify then I think it is better to implement `TryFrom` trait

File: docs/network-setup.md (hunk 8 above)
Comment (dianpopa) [reply docs/network-setup.md:82]: @vibharya you mean let them know that the driver will assign one for them?

As far as the CHANGELOG, yes, we need to remove the entries you added.

File: src/utils/src/net/mac.rs (hunk 6 above)
Comment (dianpopa) [reply src/utils/src/net/mac.rs]: This was meant as an interface, i say we leave it as it is

File: docs/network-setup.md (hunk 8 above)
Comment (vibharya) [reply docs/network-setup.md:82]: Yes, let them know one will be assigned for them. And that we cannot retrieve/report it using FC.

Yes, we can remove the entries I added and instead mention that they are now reported as null (or "None" or excluded - not sure exactly what/vm/config is returning now in this case) pre and post snapshot resume.
Hence fixing [this part](https://github.com/firecracker-microvm/firecracker/issues/2943#:~:text=net%20devices%20MAC%20addresses%20are%20reported%20as%20all%20zeros%20if%20left%20unconfigured.%20View%20Partial%20fix%20for%20restore%20snapshot%20GET%20/vm/config%20bug%C2%A0%232806%20(comment)) of the issue.

Comment (ShadowCurse) [reply docs/network-setup.md:82]: Btw, can't we retrieve the MAC with `write_config`?

Comment (ShadowCurse) [reply docs/network-setup.md:82]: @dianpopa @vibharya I changed `CHANGELOG`. Your thoughts?

File: src/vmm/src/persist.rs [hunk 10]
```diff
@@ -277,6 +250,7 @@ fn snapshot_state_to_file(
     let mut snapshot_file = OpenOptions::new()
         .create(true)
         .write(true)
+        .truncate(true)
```
[the remark this answers is not in the document]
Comment (pb8o) [reply src/vmm/src/persist.rs:253]: Maybe add more detail to the body of the commit, what happens if this is not set.

**Review by pb8o — DISMISSED** [review]

File: docs/network-setup.md (hunk 8 above)
Comment (dianpopa) [reply docs/network-setup.md:82]: sounds good to me

Comment (vibharya) [reply docs/network-setup.md:82]: Have you been able to retrieve the MAC address using `write_config`?

Also, what do we get when the MAC address is left unspecified currently?

File: src/devices/src/virtio/net/persist.rs [hunk 11]
```diff
@@ -23,10 +24,39 @@ use super::{NUM_QUEUES, QUEUE_SIZE};
 use crate::virtio::persist::{Error as VirtioStateError, VirtioDeviceState};
 use crate::virtio::{DeviceState, TYPE_NET};
 
-#[derive(Clone, Versionize)]
+#[derive(Debug, Default, Clone, Versionize)]
 // NOTICE: Any changes to this structure require a snapshot version bump.
 pub struct NetConfigSpaceState {
-    guest_mac: [u8; MAC_ADDR_LEN],
+    #[version(end = 2, default_fn = "def_guest_mac_old")]
+    guest_mac_old: MacAddr,
```
Comment (dianpopa) [inline src/devices/src/virtio/net/persist.rs:31]: I am actually surprised that the deserialization still works after changing the type of `guest_mac_old`. 
I guess this is due to how bincode does the serialization/deserialization. I suppose that `MacAddr` which behind the scenes is a struct with only one field will get deserialized to exactly what that one field contains.
So that 
```
MacAddr {byest [u8; MAC_ADDR_LEN] 
```
is equivalent to `[u8; MAC_ADDR_LEN]`
To be honest, I do not feel comfortable with this change since this implies that we will depend on how bincode works  (which can very well change in future). 
Also, I feel that when doing such changes in structs that get versionized, we should keep the changes as suggestive  as possible (history wise) by keeping old fields exactly as they were before the addition and apply the freedom of naming and type only on new fields.
In  a nutshell, the way I would have done it:

```
NetConfigSpaceState {
guest_mac: [u8; MAC_ADDR_LEN], -> exaclty as before,
guest_mac_opt/guest_mac_new: Option<MacAddr>,
}
```

Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs:31]: Well, in my opinion, `MacAddr` should not even be a struct (well it will be) but just: `pub struct`

File: docs/network-setup.md (hunk 8 above)
Comment (dianpopa) [reply docs/network-setup.md:82]: When the MAC address is left unspecified, we get back `None`. Exactly what the user configured.

In oder to receive the MAC_ADDR from the driver through `write_config`, we need to negotiate the `VIRTIO_NET_F_CTRL_MAC_ADDR` which needs to have a control queue set up. However, since our initial intent with the `GetConfig` was to give back the config provided by the user, we decided that obtaining the actual mac addr set by the driver is not part of that feature. Also, if the user has a real need to know the mac address from inside the microVM, they always have the possibility to specify one.

File: src/devices/src/virtio/net/persist.rs (hunk 11 above)
Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs:31]: Well in my opnion `MacAddr` should be:
```rust
pub struct MacAddr([u8; MAC_ADDR_LEN])
```
instead of what it is not. Then it is clear that it will always be ser/deser as just 6 bytes. And I don't think where is any way this bytes would be ser/de in any other way.

Also about changing the name to `guset_mac` and `guest_mac_opt/_new`, I think this is confusing, because in new versions of FC only `guest_mac_opt/_new` would be used. Then if someone would work with this `NetConfig` it would be unclear what field to use.

File: docs/network-setup.md (hunk 8 above)
Comment (vibharya) [reply docs/network-setup.md:82]: Just verified that the output of GET /vm/config (even for boot from snapshot) would show the network interface as something like
`{
      "iface_id":"1",
      "host_dev_name":"vmtap30",
      "guest_mac":null,
      "rx_rate_limiter":null,
      "tx_rate_limiter":null
}`
Thats good, was the case previously as well.


> When the MAC address is left unspecified, we get back `None`. Exactly what the user configured.
> 
> In oder to receive the MAC_ADDR from the driver through `write_config`, we need to negotiate the `VIRTIO_NET_F_CTRL_MAC_ADDR` which needs to have a control queue set up. However, since our initial intent with the `GetConfig` was to give back the config provided by the user, we decided that obtaining the actual mac addr set by the driver is not part of that feature. Also, if the user has a real need to know the mac address from inside the microVM, they always have the possibility to specify one.

Yes, I understand and agree with this part 100%.
I'm only wondering if we need a note in the snapshot_support doc regarding this or not.

File: src/devices/src/virtio/net/persist.rs (hunk 11 above)
Comment (dianpopa) [reply src/devices/src/virtio/net/persist.rs:31]: > Well in my opnion `MacAddr` should be:
> 
> ```rust
> pub struct MacAddr([u8; MAC_ADDR_LEN])
> ```
> 
> instead of what it is now. Then it is clear that it will always be ser/deser as just 6 bytes. And I don't think where is any way this bytes would be ser/de in any other way.
> 
> Also about changing the name to `guset_mac` and `guest_mac_opt/_new`, I think this is confusing, because in new versions of FC only `guest_mac_opt/_new` would be used. Then if someone would work with this `NetConfig` it would be unclear what field to use.

I can only agree that `MacAddr` should have been exactly that, but it s not right now, true? My personal take on this is that I would not choose to change an old field (which indeed was not really the best choice) on which snapshot compatibility depends for the sake of readability? (I guess).

I was saying in my previous comment that we should have the changes to these fields tell a story. Maybe the names suggested by me are not the happiest choice and we can go for some other options (`guest_mac_v2`
etc). Also, If we need to change the type of `guest_mac` in the future, what would we call it that time?

Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs:31]: Ok, we can have a bit ugly, but very story telling naming like `guest_mac_pre_1_2`, `guest_mac_1_2` and so on.
Also to make old fields more explicit we can use `#[deprecated]`like this:
```rust
 #[deprecated(since = "1.2.0", note = "this field is deprecated")]
 #[version(end = 2, default_fn = "def_guest_mac_old")]
 guest_mac_pre_1_2: [u8; MAC_ADDR_LEN],
 #[version(start = 2, de_fn = "de_guest_mac", ser_fn = "ser_guest_mac")]
 guest_mac_1_2: Option<MacAddr>,
```

Comment (dianpopa) [reply src/devices/src/virtio/net/persist.rs:31]: Indeed it is a bit ugly to add all the number transitions; 
I think that end = 2, start  = 2  suffice to correlate to the version map.
The deprecated thingy I like a lot, since it s a way to also correlate the field to the FC version.

Comment (ShadowCurse) [reply src/devices/src/virtio/net/persist.rs]: Well, lets leave adding deprecation to our future selfs. It deserves a separate PR because there are several places where we updated fields this way.

**Review by dianpopa — DISMISSED** [review]

**Review by vibharya — APPROVED** [review]
LGTM

**Review by roypat — APPROVED** [review]
